File: kora-app-annotation-processor/src/main/java/ru/tinkoff/kora/kora/app/annotation/processor/interceptor/ComponentInterceptors.java
package ru.tinkoff.kora.kora.app.annotation.processor.interceptor;

import ru.tinkoff.kora.kora.app.annotation.processor.ProcessingContext;
import ru.tinkoff.kora.kora.app.annotation.processor.component.ResolvedComponent;

import javax.annotation.Nullable;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

public class ComponentInterceptors {
    private final ProcessingContext ctx;
    private final List<ComponentInterceptor> interceptors;

    private ComponentInterceptors(ProcessingContext ctx, List<ComponentInterceptor> interceptors) {
        this.ctx = ctx;
        this.interceptors = interceptors;
    }

    public static ComponentInterceptors parseInterceptors(ProcessingContext ctx, Collection<ResolvedComponent> components) {
        var interceptors = new ArrayList<ComponentInterceptor>();
        for (var component : components) {
            var factory = parseInterceptor(ctx, component);
            if (factory != null) {
                interceptors.add(factory);
            }
        }
        return new ComponentInterceptors(ctx, interceptors);
    }

    @Nullable
    private static ComponentInterceptor parseInterceptor(ProcessingContext ctx, ResolvedComponent component) {
        if (!ctx.serviceTypeHelper.isInterceptor(component.type())) {
            return null;
        }
        var interceptedType = ctx.serviceTypeHelper.getInterceptedType(component.type());
        return new ComponentInterceptor(component, interceptedType);
    }

    public List<ComponentInterceptor> interceptorsFor(ResolvedComponent component) {
        var type = component.type();
        return this.interceptors.stream()
            .filter(interceptor -> this.ctx.types.isSameType(type, interceptor.interceptType()))
            .filter(interceptor -> {
                // todo tag matches
                return interceptor.component().tags().equals(component.tags());
            })
            .toList();
    }
}

File: kora-app-symbol-processor/src/main/kotlin/ru/tinkoff/kora/kora/app/ksp/interceptor/ComponentInterceptors.kt
package ru.tinkoff.kora.kora.app.ksp.interceptor

import com.google.devtools.ksp.processing.Resolver
import ru.tinkoff.kora.kora.app.ksp.ProcessingContext
import ru.tinkoff.kora.kora.app.ksp.component.ResolvedComponent
import ru.tinkoff.kora.kora.app.ksp.declaration.ComponentDeclaration

data class ComponentInterceptors(
    private val resolver: Resolver,
    private val interceptors: MutableList<ComponentInterceptor> = mutableListOf()
) {

    companion object {
        fun parseInterceptors(ctx: ProcessingContext, components: Collection<ResolvedComponent>): ComponentInterceptors {
            val interceptors = ArrayList<ComponentInterceptor>()
            for (component in components) {
                val factory = parseInterceptor(ctx, component)
                if (factory != null) {
                    interceptors.add(factory)
                }
            }
            return ComponentInterceptors(ctx.resolver, interceptors)
        }

        private fun parseInterceptor(ctx: ProcessingContext, component: ResolvedComponent): ComponentInterceptor? {
            if (!ctx.serviceTypesHelper.isInterceptor(component.type)) {
                return null
            }
            val interceptType = ctx.serviceTypesHelper.interceptType(component.type)
            return ComponentInterceptor(component, component.declaration, interceptType)
        }
    }

    fun interceptorsFor(descriptor: ComponentDeclaration): List<ComponentInterceptor> {
        val type = descriptor.type.makeNotNullable()

        return this.interceptors.filter { interceptor ->
            val realInterceptorType = interceptor.interceptType.makeNotNullable()
            type == realInterceptorType && descriptor.tags.containsAll(interceptor.declaration.tags)
        }
    }
}
